### ISSUE TITLE
Cannot assign a device for operation global_tensor_mem_usage_deviceGPU1/MaxBytesInUse

### ISSUE THREAD

**Issue by Shilpil** [rwth-i6/returnn-experiments#28]
We are trying to run Librispeech corpus with the latest base2.conv2l.specaug.curric3 config file.
It used to take 0.5 hrs for 1 epoch for 1 Tesla V100 GPU so we upgraded to 4 GPUs. Suddenly it started throwing this error:

```
Caused by op 'global_tensor_mem_usage_deviceGPU1/MaxBytesInUse', defined at:
  File "./rnn.py", line 655, in <module>
    main(sys.argv)
  File "./rnn.py", line 643, in main
    execute_main_task()
  File "./rnn.py", line 453, in execute_main_task
    engine.train()
  File "/home/ubuntu/returnn/TFEngine.py", line 1213, in train
    self.train_epoch()
  File "/home/ubuntu/returnn/TFEngine.py", line 1321, in train_epoch
    trainer.run(report_prefix=("pre" if self.is_pretrain_epoch() else "") + "train epoch %s" % self.epoch)
  File "/home/ubuntu/returnn/TFEngine.py", line 522, in run
    fetches_dict = self._get_fetches_dict()
  File "/home/ubuntu/returnn/TFEngine.py", line 133, in _get_fetches_dict
    with_summary=True, with_size=True)
  File "/home/ubuntu/returnn/TFNetwork.py", line 1027, in get_fetches_dict
    d["mem_usage:%s" % os.path.basename(dev.name.replace("/device:", "/"))] = TFUtil.mem_usage_for_dev(dev.name)
  File "/home/ubuntu/returnn/TFUtil.py", line 7752, in mem_usage_for_dev
    return global_tensor(get, "mem_usage_%s" % scope_name)
  File "/home/ubuntu/returnn/TFUtil.py", line 5805, in global_tensor
    v = f()
  File "/home/ubuntu/returnn/TFUtil.py", line 7748, in get
    return bytes_in_use()
  File "/home/ubuntu/returnn/.rwth/lib/python3.5/site-packages/tensorflow/contrib/memory_stats/python/ops/memory_stats_ops.py", line 41, in MaxBytesInUse
    return gen_memory_stats_ops.max_bytes_in_use()
  File "/home/ubuntu/returnn/.rwth/lib/python3.5/site-packages/tensorflow/contrib/memory_stats/ops/gen_memory_stats_ops.py", line 214, in max_bytes_in_use
    "MaxBytesInUse", name=name)
  File "/home/ubuntu/returnn/.rwth/lib/python3.5/site-packages/tensorflow/python/framework/op_def_library.py", line 788, in _apply_op_helper
    op_def=op_def)
  File "/home/ubuntu/returnn/.rwth/lib/python3.5/site-packages/tensorflow/python/util/deprecation.py", line 507, in new_func
    return func(*args, **kwargs)
  File "/home/ubuntu/returnn/.rwth/lib/python3.5/site-packages/tensorflow/python/framework/ops.py", line 3300, in create_op
    op_def=op_def)
  File "/home/ubuntu/returnn/.rwth/lib/python3.5/site-packages/tensorflow/python/framework/ops.py", line 1801, in __init__
    self._traceback = tf_stack.extract_stack()

InvalidArgumentError (see above for traceback): Cannot assign a device for operation global_tensor_mem_usage_deviceGPU1/MaxBytesInUse: node global_tensor_mem_usage_deviceGPU1/MaxBytesInUse (defined at /home/ubuntu/returnn/TFUtil.py:7748) was explicitly assigned to /device:GPU:1 but available devices are [ /job:localhost/replica:0/task:0/device:CPU:0, /job:localhost/replica:0/task:0/device:GPU:0, /job:localhost/replica:0/task:0/device:XLA_CPU:0, /job:localhost/replica:0/task:0/device:XLA_GPU:0, /job:localhost/replica:0/task:0/device:XLA_GPU:1, /job:localhost/replica:0/task:0/device:XLA_GPU:2, /job:localhost/replica:0/task:0/device:XLA_GPU:3 ]. Make sure the device specification refers to a valid device.
         [[node global_tensor_mem_usage_deviceGPU1/MaxBytesInUse (defined at /home/ubuntu/returnn/TFUtil.py:7748) ]]
```

We tried debugging by running basic commands like nvidia-smi and it seems fine:

``` 
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 418.87.01    Driver Version: 418.87.01    CUDA Version: 10.1     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+===============================|
|   0  Tesla V100-SXM2...  Off  | 00000000:00:1B.0 Off |                    0 |
| N/A   45C    P0    52W / 300W |      0MiB / 16130MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   1  Tesla V100-SXM2...  Off  | 00000000:00:1C.0 Off |                    0 |
| N/A   47C    P0    52W / 300W |      0MiB / 16130MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   2  Tesla V100-SXM2...  Off  | 00000000:00:1D.0 Off |                    0 |
| N/A   45C    P0    54W / 300W |      0MiB / 16130MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   3  Tesla V100-SXM2...  Off  | 00000000:00:1E.0 Off |                    0 |
| N/A   46C    P0    57W / 300W |      0MiB / 16130MiB |      4%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=========================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+
```

We are using tensorflow-gpu (1.13.2) and Cuda V7.5.17. 

I will be thankful for any help regarding this issue.Please help.

**Comment by albertz** [comment]
@pavelgolik I remember you had a similar problem, and solved it somehow? But you did not pushed any fixes to the code?

**Comment by pavelgolik** [comment]
I'm not sure if this is the same error - the one I reported was related to the XLA devices. My (temporary) workaround was to filter them out by adding `if "XLA" in dev.device_type: continue` in the beginning of the for loop in https://github.com/rwth-i6/returnn/blob/master/TFUtil.py#L3382

Anyway, this error is related to `MaxBytesInUse`. @Shilpil, are you setting `CUDA_VISIBLE_DEVICES` explicitly?

**Comment by Shilpil** [comment]
Yes we also tried that by running it as : 
CUDA_VISIBLE_DEVICES = 0,1,2,3 ./rnn.py 2019-librispeech-system/attention/base2.conv2l.specaug.curric3.config

**Comment by pavelgolik** [comment]
Another possible workaround was to set `tf_log_memory_usage = False`.

**Comment by albertz** [comment]
The error:
`InvalidArgumentError (see above for traceback): Cannot assign a device for operation global_tensor_mem_usage_deviceGPU1/MaxBytesInUse: node global_tensor_mem_usage_deviceGPU1/MaxBytesInUse (defined at /home/ubuntu/returnn/TFUtil.py:7748) was explicitly assigned to /device:GPU:1 but available devices are [ /job:localhost/replica:0/task:0/device:CPU:0, /job:localhost/replica:0/task:0/device:GPU:0, /job:localhost/replica:0/task:0/device:XLA_CPU:0, /job:localhost/replica:0/task:0/device:XLA_GPU:0, /job:localhost/replica:0/task:0/device:XLA_GPU:1, /job:localhost/replica:0/task:0/device:XLA_GPU:2, /job:localhost/replica:0/task:0/device:XLA_GPU:3 ]. Make sure the device specification refers to a valid device.`

So I wonder, why does TF has GPU:0 & XLA_GPU:0 available, but not GPU:1, but only XLA_GPU:1?

**Comment by albertz** [comment]
I found [TF #30748](https://github.com/tensorflow/tensorflow/issues/30748), where one Google dev states:
> XLA creates an XLA_GPU device for every present on the system whereas TF creates a GPU device only for GPUs suitable for compute (i.e. ignores slower GPUs) so that could explain what you're seeing. TF logs out "Ignoring visible gpu device" when it does this enhanced filtering so you should see it in your logs.

Btw, anyway, is this for multi-GPU training? How are you actually doing this? Just making all the GPUs available to Returnn will not be helpful at all (how should it?). It will not use them.

Please check [our documentation about multi-GPU training](https://returnn.readthedocs.io/en/latest/multi_gpu.html) about how to actually do multi-GPU training.

**Comment by Shilpil** [comment]
Hi,
I installed Horovod and tested it by running mpirun -np 8 -mca pml ob1 -mca btl ^openib python3 returnn/demos/demo-horovod-mpi.py

But when I run the librispeech config file I get this error using the following command :

mpirun -np 4 -H localhost:4 -bind-to none -map-by slot -x NCCL_DEBUG=INFO -x LD_LIBRARY_PATH -x PATH -mca pml ob1 -mca btl ^openib python rnn.py 2019-librispeech-system/attention/base2.conv2l.specaug.curric3.config ++use_horovod 1 

I get this error:

InvalidArgumentError (see above for traceback): Cannot assign a device for operation GetDeviceAttr_1: node GetDeviceAttr_1 (defined at <string>:67) was explicitly assigned to /job:localhost/replica:0/task:0/device:GPU:1 but available devices are [ /job:localhost/replica:0/task:0/device:CPU:0, /job:localhost/replica:0/task:0/device:GPU:0, /job:localhost/replica:0/task:0/device:XLA_CPU:0, /job:localhost/replica:0/task:0/device:XLA_GPU:1 ]. Make sure the device specification refers to a valid device.
         [[node GetDeviceAttr_1 (defined at <string>:67) ]]

**Comment by albertz** [comment]
Can you post the full stdout somewhere?

The idea in Horovod is that for N GPUs (either on a single host, or even multiple hosts), it will start N instances of the program (or Returnn), and each single instance only has access to one single GPU, and all the instances communicate with each other through special Horovod TF ops. So in your output, it seems like there is one instance of Returnn which actually sees two GPUs (GPU:0 and XLA_GPU:1).

@pavelgolik Maybe you have some idea how this happens?

**Comment by kazuki-irie** [comment]
CC'ing @Gerstenberger, in case you might also have some idea.

**Comment by Shilpil** [comment]
Hi,
I am attaching the full output.
[output.txt](https://github.com/rwth-i6/returnn-experiments/files/3909658/output.txt)

I have a small query,
If we upgrade the instance on AWS from 1 GPU to 4 GPUs do we need to reinstall Horovod?
Can this be the reason the error is thrown?

Thank you all for helping me with this issue.

**Comment by pavelgolik** [comment]
Please try the work-around mentioned in https://github.com/rwth-i6/returnn-experiments/issues/28#issuecomment-557051791

> If we upgrade the instance on AWS from 1 GPU to 4 GPUs do we need to reinstall Horovod?

No, I don't think so.

**Comment by albertz** [comment]
> I am attaching the full output.
> [output.txt](https://github.com/rwth-i6/returnn-experiments/files/3909658/output.txt)

Thanks.

Some notes:

* `++use_horovod 4`: You don't need 4 here, this is just a True/False flag.
* `horovod_reduce_type = "param"` is probably more efficient (see the [multi GPU documentation](https://returnn.readthedocs.io/en/latest/multi_gpu.html)).
* You use TF 1.13.2. Did you try another version? E.g. TF 1.14? Or maybe also older.
* In the function `get_device_attr`, there is `dev = dev.replace(":XLA_", ":")`. When you remove that, does it work then? If not, can you post the output log for that as well?

**Comment by Shilpil** [comment]
Using comment https://github.com/rwth-i6/returnn-experiments/issues/28#issuecomment-557051791 worked. Thank you. It started running after I put `if "XLA" in dev.device_type: continue` in the beginning of the for loop in function` get_tf_list_local_devices`.
But I could not see the acceleration in the timing. Is there anything I am missing?

Here is the command I used for 4 GPUs : 
mpirun -np 4 -bind-to none -map-by slot -x NCCL_DEBUG=INFO -x LD_LIBRARY_PATH -x PATH -x HOROVOD_TIMELINE -x DEBUG -mca pml ob1 -mca btl ^openib python3 rnn.py 2019-librispeech-system/attention/base2.conv2l.specaug.curric3.config ++use_horovod 1

Here is the log for 4 GPUs:
pretrain epoch 43, step 20, cost:ctc 1.3570787483922613, cost:output/output_prob 1.0764964977539009, error:ctc 0.2901119335438125, error:decision 0.0, error:output/output_prob 0.21921641280641777, loss 652.19824, max_size:classes , num_seqs 6, 1.883 sec/step, elapsed 0:01:38, exp. remaining 0:41:37, complete 3.80%

This is the log for 1 GPU:
pretrain epoch 43, step 73, cost:ctc 1.245547996649293, cost:output/output_prob 1.0171804541269012, error:ctc 0.2908366620540619, error:decision 0.0, error:output/output_prob 0.21115538477897644, loss 567.9448, max_size:classes 45, max_size:data 1626, num_seqs 6, 0.998 sec/step, elapsed 0:01:37, exp. remaining 0:41:00, complete 3.80%

In both of them it says 3.8 % complete and remaining time is around 41 minutes.
Please help me if I am going wrong in any step.

I checked the usage of the GPUs using nvidia-smi command and all the 4 GPUs were being used to the max extent. No other programs were running in parallel.

**Comment by albertz** [comment]
Please [check the documentation about multi-GPU training](https://returnn.readthedocs.io/en/latest/multi_gpu.html). I guess (but I cannot tell from your output) that this is a bottleneck in the dataset. Did you follow the documentation? Relevant for this is to check this part in the log:
```
train epoch 1, finished after 2941 steps, 0:28:58 elapsed (99.3% computing time)
```
Look at the computing time in particular. That numbers measures how much relative time was spend inside TF session.run. If this is below 90% or so, it means that you wasted some time elsewhere, e.g. the dataset loading.

If that is the case, you can try to use `HDFDataset` instead. There is the tool `hdf_dump.py` to convert your dataset into a HDF dataset.

Otherwise, what `horovod_reduce_type` do you use know? And what `horovod_param_sync_step`? These are all very relevant (please check the documentation).

**Comment by albertz** [comment]
Closing this now.
Please open a new issue in the Returnn repo (not in this experiments repo) if there are still problems, although I think this is fixed, or not?
